elftools/elf/segments.py

```diff
@@ -119,11 +120,20 @@ def iter_notes(self):
             self.stream.seek(offset)
             # n_namesz is 4-byte aligned.
             disk_namesz = roundup(note['n_namesz'], 2)
+            # skip parsing truncated notes.
+            if offset+disk_namesz+roundup(note['n_descsz'], 2) > end:
+                break;
             note['n_name'] = bytes2str(
                 CString('').parse(self.stream.read(disk_namesz)))
             offset += disk_namesz
 
             desc_data = bytes2str(self.stream.read(note['n_descsz']))
+            if note['n_name'] == 'GNU':                                       
```

Comment:
I'm not actually sure what you are asking here.
IF you are asking why we have to do a linear walk of a dick to do a lookup by value.. then AFAIK that jsut how dicts and hash tables are implemented. Somebody is having to do this anyway, in elftools you seem to be letting the helper lib do it.

if you are asking why we need to only do this when the name is GNU. well because its the correct way. This Ids are not unique they are only sub types of the note name. i.e. They can only be used if the Note name is GNU, there are other tools, vendors, oems that add notes that aren't GNU and they also start at id 0 for the type. As such the current code incorrectly names these notes as for example note-id.  Which means that with some elf files, elftools returns the wrong content, e.g. a build-id that is not actually a build-id but rather some random note content.